File: LeetCode/3011-find-if-array-can-be-sorted/3011-find-if-array-can-be-sorted.java
class Solution {
    public boolean canSortArray(int[] nums) {
        int numOfSetBits = Integer.bitCount(nums[0]);
        int maxOfSegment = nums[0];
        int minOfSegment = nums[0];
        int maxOfPrevSegment = Integer.MIN_VALUE;

        for (int i = 1; i < nums.length; i++) {
            if (Integer.bitCount(nums[i]) == numOfSetBits) {
                maxOfSegment = Math.max(maxOfSegment, nums[i]);
                minOfSegment = Math.min(minOfSegment, nums[i]);
            } else {
                if (minOfSegment < maxOfPrevSegment) {
                    return false;
                }
                maxOfPrevSegment = maxOfSegment;
                maxOfSegment = nums[i];
                minOfSegment = nums[i];
                numOfSetBits = Integer.bitCount(nums[i]);
            }
        }
        
        return minOfSegment >= maxOfPrevSegment;
    }
}





// class Solution {
//     public boolean canSortArray(int[] nums) {
//         int n = nums.length;
//         boolean swapped = true;
//         for (int i = 0; i < n; i++) {
//             swapped = false;
//             for (int j = 0; j < n - i - 1; j++) {
//                 if (nums[j] <= nums[j + 1]) {
//                     continue;
//                 } else {
//                     if (Integer.bitCount(nums[j]) == Integer.bitCount(nums[j + 1])) {
//                         int temp = nums[j];
//                         nums[j] = nums[j + 1];
//                         nums[j + 1] = temp;
//                         swapped = true;
//                     } else {
//                         return false;
//                     }
//                 }
//             }
//             if (!swapped) {
//                 break;
//             }
//         }
//         return true;
//     }
// }

File: LeetCode/3011-find-if-array-can-be-sorted/3011-find-if-array-can-be-sorted.py
class Solution:
    def canSortArray(self, nums: List[int]) -> bool:
        numOfSetBits = bin(nums[0]).count('1')
        maxOfSegment = nums[0]
        minOfSegment = nums[0]
        maxOfPrevSegment = float('-inf')

        for i in range(1, len(nums)):
            if bin(nums[i]).count('1') == numOfSetBits:
                maxOfSegment = max(maxOfSegment, nums[i])
                minOfSegment = min(minOfSegment, nums[i])
            else:
                if minOfSegment < maxOfPrevSegment:
                    return False
                maxOfPrevSegment = maxOfSegment
                maxOfSegment = nums[i]
                minOfSegment = nums[i]
                numOfSetBits = bin(nums[i]).count('1')
        
        return minOfSegment >= maxOfPrevSegment





# class Solution:
#     def canSortArray(self, nums: List[int]) -> bool:
#         n = len(nums)
#         swapped = True
#         for i in range(n):
#             swapped = False
#             for j in range(n - i - 1):
#                 if nums[j] <= nums[j + 1]:
#                     continue
#                 else:
#                     if bin(nums[j]).count('1') == bin(nums[j + 1]).count('1'):
#                         nums[j], nums[j + 1] = nums[j + 1], nums[j]
#                         swapped = True
#                     else:
#                         return False
#             if not swapped:
#                 break
#         return True